File: NOVEMBER-2025/20-11-2025/757.erl
-spec intersection_size_two([[integer()]]) -> integer().
intersection_size_two(Intervals) ->
    Sorted = lists:sort(
        fun([_, E1], [_, E2]) -> E1 =< E2 end,
        Intervals
    ),

    [Start0, End0] = hd(Sorted),
    Prev1 = End0 - 1,
    Prev2 = End0,
    Count0 = 2,

    {_, _, FinalCount} =
        lists:foldl(
            fun([Start, End], {P1, P2, C}) ->
                if
                    P2 < Start ->
                        {End - 1, End, C + 2};

                    P1 < Start ->
                        NewP1 =
                            case End =:= P2 of
                                true  -> End - 1;
                                false -> End
                            end,

                        {FinalP1, FinalP2} =
                            case NewP1 > P2 of
                                true  -> {P2, NewP1};
                                false -> {NewP1, P2}
                            end,

                        {FinalP1, FinalP2, C + 1};

                    true ->
                        {P1, P2, C}
                end
            end,
            {Prev1, Prev2, Count0},
            tl(Sorted)
        ),

    FinalCount.

File: NOVEMBER-2025/20-11-2025/757.ex
defmodule Solution do
  @spec intersection_size_two(intervals :: [[integer]]) :: integer
  def intersection_size_two(intervals) do
    intervals =
      Enum.sort(intervals, fn [_, e1], [_, e2] ->
        e1 <= e2
      end)

    [start0, end0] = hd(intervals)
    prev1 = end0 - 1
    prev2 = end0
    count = 2

    {_, _, result} =
      intervals
      |> Enum.drop(1)
      |> Enum.reduce({prev1, prev2, count}, fn [start, endi], {p1, p2, c} ->
        cond do
          p2 < start ->
            {endi - 1, endi, c + 2}

          p1 < start ->
            new_p1 =
              if endi == p2 do
                endi - 1
              else
                endi
              end
            {p1_new, p2_new} =
              if new_p1 > p2 do
                {p2, new_p1}
              else
                {new_p1, p2}
              end

            {p1_new, p2_new, c + 1}
          true ->
            {p1, p2, c}
        end
      end)

    result
  end
end